src/Bitcoin/SegwitAddress.cpp

```diff
@@ -14,37 +14,20 @@
 using namespace TW::Bitcoin;
 
 bool SegwitAddress::isValid(const std::string& string) {
-    auto dec = Bech32::decode(string);
-    if (dec.second.empty()) {
-        return false;
-    }
-
-    Data conv;
-    if (!Bech32::convertBits<5, 8, false>(conv, Data(dec.second.begin() + 1, dec.second.end())) ||
-        conv.size() < 2 || conv.size() > 40 || dec.second[0] > 16 ||
-        (dec.second[0] == 0 && conv.size() != 20 && conv.size() != 32)) {
-        return false;
-    }
-
-    return true;
+    return decode(string).second;
 }
 
 bool SegwitAddress::isValid(const std::string& string, const std::string& hrp) {
-    auto dec = Bech32::decode(string);
-    if (dec.second.empty()) {
+    auto decoded = decode(string);
+    if (!decoded.second) {
         return false;
     }
+    // extra step to check hrp
+    auto dec = Bech32::decode(string);
```

Comment:
add an optional hrp parameter to `SegwitAddress::decode` to avoid decoding twice

Reply:
done